http/cookies.lua

```diff
@@ -0,0 +1,324 @@
+	assert(data.value, "cookie must contain a `value` field")
+	local cookie = {tostring(data.key) .. "=" .. tostring(data.value)}
+	if data.expires then
+		cookie[#cookie + 1] = "Expires=" .. util.imf_date(data.expires)
+	end
+	if data.max_age then
+		cookie[#cookie + 1] = "Max-Age=" .. tostring(math.floor(data.max_age))
+	end
+	if data.domain then
+		cookie[#cookie + 1] = "Domain=" .. data.domain
+	end
+	if data.path then
+		cookie[#cookie + 1] = "Path=" .. data.path
+	end
+	if data.secure then
+		cookie[#cookie + 1] = "Secure"
+	end
+	if data.http_only then
+		cookie[#cookie + 1] = "HttpOnly"
+	end
+	if data.same_site then
+		if data.same_site == true then
+			data.same_site = "Strict"
+		end
+		cookie[#cookie + 1] = "SameSite=" .. data.same_site:gsub("^.",
+			string.upper)
+	end
+	return table.concat(cookie, "; ")
+end
+
+local function iterate_cookies(cookie)
+	return pairs(assert(http_patts.Cookie:match(cookie)))
+end
+
+local function parse_cookies(cookie)
+	local cookies = {}
+	for k, v in iterate_cookies(cookie) do
+		cookies[k] = v
+		cookies[#cookies + 1] = {k, v}
+	end
+	table.sort(cookies, function(t1, t2)
+		return t1[1] < t2[1]
+	end)
+	return cookies
+end
+
+local cookiejar_methods = {}
+local cookiejar_mt = {__index = cookiejar_methods}
+
+local function new_cookiejar(psl_object)
+	psl_object = psl_object or psl.builtin()
+	return setmetatable({cookies={}, psl = psl_object}, cookiejar_mt)
+end
+
+function cookiejar_methods:add(cookie)
+	cookie.last_access = os.time()
+	local domain, path, key = cookie.domain, cookie.path, cookie.key
+	local cookies = self.cookies
+	if cookies[domain] and cookies[domain][path] then
+		local old_cookie = cookies[domain][path][key]
+		if old_cookie then
+			cookie.creation = old_cookie.creation
+		end
+	end
+
+	local old_cookie = self:get_by_indexname(cookie.indexname)
+	if old_cookie then
+		self:remove_cookie(old_cookie)
+	end
+	if cookie.persistent then
+		local cookie_exp_time = cookie.expires
+		local inserted = false
+		for i=1, #cookies do
+			-- insert into first spot where cookie expires after
+			if cookies[i].expires < cookie_exp_time then
+				inserted = true
+				table.insert(cookies, i, cookie)
+			end
+		end
+		if not inserted then
+			cookies[#cookies + 1] = cookie
+		end
+	else
+		cookie.expires = math.huge
+		table.insert(cookies, 1, cookie)
+	end
+
+	if not cookies[domain] then
+		cookies[domain] = {
+			[path] = {
+				[key] = cookie
+			}
+		}
+	elseif not cookies[domain][path] then
+		cookies[domain][path] = {
+			[key] = cookie
+		}
+	else
+		cookies[domain][path][key] = cookie
+	end
+end
+
+function cookiejar_methods:get_expired(time)
+	time = time or os.time()
+	local cookies = self.cookies
+	local returned_cookies = {}
+	for i=#cookies, 1, -1 do
+		local cookie = cookies[i]
+		if cookie.expires > time then
+			break
+		end
+		returned_cookies[#returned_cookies + 1] = cookie
+	end
+	return returned_cookies
+end
+
+function cookiejar_methods:get_by_indexname(name)
```

Comment:
This is hugely expensive splitting it apart, you only need a `:get(domain, path, key)`